src/NHibernate.Test/TestCase.cs

```diff
@@ -221,13 +222,26 @@ private bool CheckSessionsWereClosed()
 			var allClosed = true;
 			foreach (var session in _openedSessions)
 			{
-				session.GetSessionImplementation().TransactionContext?.WaitOne();
-				if (session.IsOpen)
+				var txContext = session.GetSessionImplementation().TransactionContext;
+				txContext?.WaitOne();
+				if (!session.IsOpen)
+					continue;
+
+				if (txContext?.ShouldCloseSessionOnDistributedTransactionCompleted ?? false)
 				{
-					log.Error($"Test case didn't close session {session.GetSessionImplementation().SessionId}, closing");
-					allClosed = false;
-					session.Close();
+					// Delayed rollback not having lock from prepare phase? Give it a chance to complete.
+					Thread.Sleep(100);
```

Comment:
Crutch for delayed disposal. Only possible fix IMO: cease delaying session disposal when an ambient transaction is ongoing.